crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2928,6 +2883,59 @@ impl<'db> KnownClass {
         matches!(self, Self::Bool)
     }
 
+    /// Determine whether instances of this class are always truthy, always falsy,
+    /// or have an ambiguous truthiness.
+    const fn bool(self) -> Truthiness {
+        match self {
+            // N.B. It's only generally safe to infer `Truthiness::AlwaysTrue` for a `KnownClass`
+            // variant if the class's `__bool__` method always returns the same thing *and* the
+            // class is `@final`.
+            //
+            // E.g. `ModuleType.__bool__` always returns `True`, but `ModuleType` is not `@final`.
```

Comment:
If e.g. `ModuleType` actually had a `def __bool__(self) -> Literal[True]` annotation, then we could rely on Liskov to say that all modules must be boolean true. But it doesn't (which is probably reasonable.

Reply:
Agree